### PR TITLE
Weapon revamp

### PR DESCRIPTION
**Pull request by Qustinnus**
yeah i'll update this later but basically refactors weapons and makes it less click spam and less easy to abuse

cl by TMTIME

:cl: Floyd + TMTIME
add: Weapon circuits are now weapon chips
del: Ship weapons no longer use cells. No more OP cannons
tweak: Phase cannons now fire three shot bursts
tweak: Explosions do less damage to the ship
balance: Seriously nerfed the fire rate of ship weapons
balance: Weapon hardpoints (Replaces machine frames) are now very expensive
balance: Ion chips require more gold
imageadd: Fancy new Kmc cannon sprites
imagedel: Old cannon sprites
/:cl:

Weapon hardpoints require 25 plasteel and 5 diamonds

### CODE REVIEW

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/obj/machinery/power/shipweapon //PHYSICAL WEAPON
+	name = "phase cannon"
+	desc = "A powerful weapon designed to take down shields.\n<span class='notice'>Alt-click to rotate it clockwise.</span>"
+	icon = 'icons/obj/96x96.dmi'
+	icon_state = "phase_cannon"
+	pixel_x = -32
+	pixel_y = -32
```
Comment (TMTIME) [inline code/modules/ftl/machinery/phase_cannon.dm]: -23 looks a little cooler but I feel this leads to an argument about consistent 2d style
dreamseeker_2018-05-14_20-26-01

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	density = 1
+
+	var/charge_rate = 200
+	var/current_charge = 0
+
+	use_power = ACTIVE_POWER_USE
+	idle_power_usage = 1000
+	active_power_usage = 20000
+
+	var/obj/item/weapon_chip/chip
```
Comment (TMTIME) [inline code/modules/ftl/machinery/phase_cannon.dm]: Needs versions with a chip in at roundstart for mapping

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			chip = null
+			current_charge = 0
+	else if(istype(W, /obj/item/weapon_chip))
+		W.loc = src
+		chip = W
+
+
+/obj/machinery/power/shipweapon/update_icon()
+	if(can_fire())
+		icon_state = "[chip.weapon_name]_fire"
```
Comment (TMTIME) [inline code/modules/ftl/machinery/phase_cannon.dm]: icon_name here

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		name = "circuit board (Phase Cannon)"
+		build_path = /obj/machinery/power/shipweapon
+
+/obj/machinery/power/shipweapon/process()
+	if(stat & (BROKEN|MAINT))
+		return
+	if(!chip)
+		current_charge = 0
+	if(!active_power_usage || avail(active_power_usage)) //Is there enough power available
+		var/load = min((chip.charge_to_fire - current_charge), charge_rate)		// charge at set rate, limited to spare capacity
```
Comment (TMTIME) [inline code/modules/ftl/machinery/phase_cannon.dm]: Currently runtimes if there is no chip installed

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+/obj/machinery/power/shipweapon/proc/can_fire()
+	return current_charge >= chip.charge_to_fire && chip
+
+/obj/machinery/power/shipweapon/proc/attempt_fire(var/datum/ship_component/target_component,var/shooter)
+	if(!can_fire())
+		return FALSE
+	current_charge = 0
+	for(var/i in 1 to chip.attack_data.shots_fired) //Fire for the amount of time
+		spawn_projectile(target_component, shooter)
```
Comment (TMTIME) [inline code/modules/ftl/machinery/phase_cannon.dm]: Spawns multiple projectiles with no delay

**Comment by Qustinnus** [comment]
i hope to finish this up this during a weekend or I'll have to wait until summer

**Comment by Qustinnus** [comment]
can someone ping me this weekend and ill work on it

File: code/modules/ftl/machinery/phase_cannon.dm (hunk 2 above)
Comment (Qustinnus) [reply code/modules/ftl/machinery/phase_cannon.dm]: I'm okay with them inserting the chip themselves

File: code/modules/ftl/machinery/phase_cannon.dm (hunk 3 above)
Comment (Qustinnus) [reply code/modules/ftl/machinery/phase_cannon.dm]: huh

File: code/modules/ftl/machinery/phase_cannon.dm (hunk 4 above)
Comment (Qustinnus) [reply code/modules/ftl/machinery/phase_cannon.dm]: good catch

File: code/modules/ftl/machinery/phase_cannon.dm (hunk 5 above)
Comment (Qustinnus) [reply code/modules/ftl/machinery/phase_cannon.dm]: forgot about that thanks

File: code/modules/ftl/machinery/phase_cannon.dm (hunk 3 above)
Comment (TMTIME) [reply code/modules/ftl/machinery/phase_cannon.dm]: chip.weapon_name needs to be chip.icon_name 

weapon name has a space and won;t render the icon iirc

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	A.yo = 0
+	A.xo = 20
+	A.attack_data = chip.attack_data //Actual info of weapon
+	A.starting = loc
+	A.fire()
+	A.target = target_component
+
+	playsound(loc, chip.fire_sound, 50, 1)
+
+	for(var/obj/machinery/computer/ftl_weapons/C in world) //this is disgusting monster
```
Comment (TMTIME) [inline code/modules/ftl/machinery/phase_cannon.dm]: we already store ftl_weapons_consoles in globals, may as well use it there

**Comment by Qustinnus** [comment]
can someoen update the maps lolmao

**Comment by Qustinnus** [comment]
@TMTIME anything we should still change?

**Comment by TMTIME** [comment]
it all works now. although ion chips now can't be made

**Comment by TMTIME** [comment]
wait fuck wrong button

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+		if(CONSTRUCTION_STATE1) //Weld the cannon in place or decon it
+
+			if(istype(W, /obj/item/weapon/weldingtool)) //Fully decon
+				var/obj/item/weapon/weldingtool/WT = W
+				if(!WT.remove_fuel(0, user))
+					if(!WT.isOn())
+						to_chat(user, "<span class='warning'>The welding tool must be on to complete this task!</span>")
+					return
+				playsound(src.loc, W.usesound, 50, 1)
+				to_chat(user, "<span class='notice'>You start disassembling the cannon...</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "you (start/begin) to disassemble \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+				if(do_after(user, 20*W.toolspeed, target = src))
+					if(!src || !WT.isOn()) return
+					to_chat(user, "<span class='notice'>You disassemble the cannon.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "you disassemble \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+					var/obj/item/stack/sheet/plasteel/P = new (loc, 25)
+					P.add_fingerprint(user)
+					W.add_fingerprint(user)
+					qdel(src)
+
+			if(istype(W, /obj/item/weapon/wrench)) //construct
+				playsound(src.loc, W.usesound, 50, 1)
+				to_chat(user, "<span class='notice'>You start bolting the cannon to the floor...</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "you begin to bolt \the [src] to the floor."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+				if(do_after(user, 20*W.toolspeed, target = src))
+					if(!src) return
+					to_chat(user, "<span class='notice'>You bolt the cannon to the floor.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You bolt \the [src] to the floor."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+					anchored = TRUE
+					state = CONSTRUCTION_STATE2
+					W.add_fingerprint(user)
+
+
+		if(CONSTRUCTION_STATE2) //unbolt the cannon or weld it down
+
+			if(istype(W, /obj/item/weapon/wrench)) //decon
+				playsound(src.loc, W.usesound, 50, 1)
+				to_chat(user, "<span class='notice'>You start unbolting the cannon from the floor...</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You start to unbolt \the [src] from the floor."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+				if(do_after(user, 20*W.toolspeed, target = src))
+					if(!src) return
+					to_chat(user, "<span class='notice'>You unbolt the cannon from the floor.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You unbolt \the [src] from the floor."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+					anchored = FALSE
+					state = CONSTRUCTION_STATE1
+					W.add_fingerprint(user)
+
+			if(istype(W, /obj/item/weapon/weldingtool)) //Contruct
+				var/obj/item/weapon/weldingtool/WT = W
+				if(!WT.remove_fuel(0, user))
+					if(!WT.isOn())
+						to_chat(user, "<span class='warning'>The welding tool must be on to complete this task!</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "\The [WT] must be on to complete this task!"

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+					return
+				playsound(src.loc, W.usesound, 50, 1)
+				to_chat(user, "<span class='notice'>You start welding the cannon's supports in place...</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You begin to weld the supports of \the [src] into place."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+				if(do_after(user, 20*W.toolspeed, target = src))
+					if(!src || !WT.isOn()) return
+					to_chat(user, "<span class='notice'>You weld the cannons' supports in place.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You weld the supports of \the [src] into place."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+					anchored = TRUE
+					state = CONSTRUCTION_STATE3
+					W.add_fingerprint(user)
+
+		if(CONSTRUCTION_STATE3) //unweld the cannon or add focusing lenses
+			if(istype(W, /obj/item/weapon/weldingtool)) //Decon
+				var/obj/item/weapon/weldingtool/WT = W
+				if(!WT.remove_fuel(0, user))
+					if(!WT.isOn())
+						to_chat(user, "<span class='warning'>The welding tool must be on to complete this task!</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "\The [WT] must be on to complete this task!"

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+					return
+				playsound(src.loc, W.usesound, 50, 1)
+				to_chat(user, "<span class='notice'>You start unwelding the cannons supports...</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You begin to unweld the supports of \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+				if(do_after(user, 20*W.toolspeed, target = src))
+					if(!src || !WT.isOn()) return
+					to_chat(user, "<span class='notice'>You unweld the cannons supports.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You unweld the supports of \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+					anchored = FALSE
+					state = CONSTRUCTION_STATE2
+					W.add_fingerprint(user)
+
+			if(istype(W, /obj/item/stack/sheet/mineral/diamond)) //Contruct
+				var/obj/item/stack/sheet/mineral/diamond/D = W
+				if(D.use(5))
+					to_chat(user, "<span class='notice'>You install focusing lenses to the cannon.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You install \the [D] as focusing lenses for \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+					state = CONSTRUCTION_STATE4
+				else
+					to_chat(user, "<span class='warning'>You need five diamonds to install lenses!</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You require at least 5 [D] to use as focusing lenses for \the [src]!"

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+
+		if(CONSTRUCTION_STATE4) //remove lenses or add chip
+			if(istype(W, /obj/item/weapon/crowbar)) //Decon
+				to_chat(user, "<span class='notice'>You pry the focusing lenses out of the cannon.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You pry the focusing lenses out of \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+				state = CONSTRUCTION_STATE3
+
+
+			if(istype(W, /obj/item/weapon_chip)) //Contruct
+				if(!user.drop_item())
+					return
+				playsound(src.loc, 'sound/items/deconstruct.ogg', 50, 1)
+				chip = W
+				W.loc = src
+				to_chat(user, "<span class='notice'>You install the [W] into the cannon.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You install \the [W] into \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+				state = CONSTRUCTION_STATE5
+
+		if(CONSTRUCTION_STATE5) //remove chip or close cover
+			if(istype(W, /obj/item/weapon/crowbar)) //Decon
+				chip.loc = src.loc
+				to_chat(user, "<span class='notice'>You remove the [chip] out of the cannon.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You remove \the [chip] out of \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 24]
```diff
@@ -0,0 +1,267 @@
...
+				chip = null
+				state = CONSTRUCTION_STATE4
+
+			if(istype(W, /obj/item/weapon/screwdriver)) //Contruct
+				to_chat(user, "<span class='notice'>You close the cannon's maintenance hatch.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You close the maintenance hatch of \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+				name = chip.weapon_name
+				connect_to_network()
+				state = CONSTRUCTION_COMPLETED
+
+		if(CONSTRUCTION_COMPLETED) //open cover
+
+			if(istype(W, /obj/item/weapon/screwdriver)) //Decon
+				to_chat(user, "<span class='notice'>You open the cannon's maintenance hatch.</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "You open the maintenance hatch of \the [src]."

File: code/modules/ftl/machinery/phase_cannon.dm [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+
+/obj/machinery/power/shipweapon/attackby(obj/item/W, mob/user, params) //someone add this thanks -no
+	switch(state)
+		if(CONSTRUCTION_STATE1) //Weld the cannon in place or decon it
+
+			if(istype(W, /obj/item/weapon/weldingtool)) //Fully decon
+				var/obj/item/weapon/weldingtool/WT = W
+				if(!WT.remove_fuel(0, user))
+					if(!WT.isOn())
+						to_chat(user, "<span class='warning'>The welding tool must be on to complete this task!</span>")
```
Comment (MadmanMartian) [inline code/modules/ftl/machinery/phase_cannon.dm]: "\The [WT] must be on to complete this task!"

**Comment by Qustinnus** [comment]
@MadmanMartian admins.txt is always going to be full of coders and thats how it should be, anyone who creates their own server should be using SQL or manually editing the file.

**Comment by TMTIME** [comment]
dreammaker_2018-08-08_22-01-56
blue

**Comment by crossedfall** [comment]
As far as code review goes, I didn't have any major problems ~~outside of the 4,518 unnecessary line changes~~.

I QA'd this on my test server and found that everything ran as expected. I was able to freely construct and deconstruct the new cannons without any issues. I also confirmed that it compiles and runs on 512 without any issues.

My primary concerns are mostly balance related. Post-revamp combat became a massive slog-fest with all of the nerfs. It's also worth noting that this has the heavy potential to increase round times since the nerfs to cannon requirements means that extensive mining is basically a requirement now to ensure that the crew has hoards of diamonds to reasonably get through combat. With round times already higher than the average SS13 round, this might turn some people off. Further monitoring with live games will be needed.

**Pull request merged**